src/Tools/ExternalAccess/VisualDiagnostics/Internal/DocumentHotReloadDiagnosticSourceProvider.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ExternalAccess.VisualDiagnostics.Contracts;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServer.Handler;
+using Microsoft.CodeAnalysis.LanguageServer.Handler.Diagnostics;
+namespace Microsoft.CodeAnalysis.ExternalAccess.VisualDiagnostics.Internal;
+
+[Export(typeof(IDiagnosticSourceProvider)), Shared]
```

Comment:
instead of implementing the concrete provider directly in roslyn - would it make sense to expose just a wrapper around `IDiagnosticSource` / `IDiagnosticSourceProvider`

Which you would then implement in entirety on the xaml side?

Reply:
That was the idea :). Are you all this can be even smaller?

I added `IHotReloadDiagnosticSource` for `IDiagnosticsSource`. `IDiagnosticSourceProvider` is pretty tiny and we still need MEF register pass thru `IDiagnosticSourceProvider`. I do not want source name to be provided from outside. To avoid conflicts with "reserved" names.

Reply:
Okay, I made it more pass thru. We now have `IHotReloadDiagnosticSourceProvider` and `IHotReloadDiagnosticSource`. I use only one MEF extension point, `IHotReloadDiagnosticManager`.